sentry-core/src/futures.rs

```diff
@@ -0,0 +1,96 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use crate::Hub;
+
+/// A future that bind a `Hub` to its execution.
+///
+/// This activates the given hub for the duration of the inner futures `poll`
+/// method.
+#[derive(Debug)]
+pub struct SentryFuture<F> {
+    hub: Arc<Hub>,
+    future: F,
+}
+
+impl<F> SentryFuture<F> {
+    /// Creates a new bound future with a `Hub`.
+    pub fn new(hub: Arc<Hub>, future: F) -> Self {
+        Self { hub, future }
+    }
+}
+
+impl<F> Future for SentryFuture<F>
+where
+    F: Future,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let hub = self.hub.clone();
+        // https://doc.rust-lang.org/std/pin/index.html#pinning-is-structural-for-field
+        let future = unsafe { self.map_unchecked_mut(|s| &mut s.future) };
```

Comment:
At first I was thinking it would be nice to use [`futures::pin_mut`](https://docs.rs/futures/0.3.5/futures/macro.pin_mut.html), but since we don't have a dependency on `futures` yet, it's probably not worth it.

Reply:
I don‘t think its worth pulling in the dependency for that. Also, I remember I read about not being able to essentially implement future combinators without unsafe code before. But I’m surprised there is no better safe alternative apparently?